### ISSUE TITLE
op_ref: Following lint leads to unconditional recursion 

### ISSUE THREAD

**Issue by LeCyberDucky** [rust-lang/rust-clippy#7572]
Lint name: op_ref

I tried this code:

```rust
use std::ops::Mul;

#[derive(Clone, Copy)]
struct Vec3D {
    pub x: f64,
    pub y: f64,
    pub z: f64,
}

impl Mul<&Vec3D> for Vec3D {
    type Output = f64;

    fn mul(self, rhs: &Vec3D) -> Self::Output {
        self.x * rhs.x + self.y * rhs.y + self.z * rhs.z
    }
}

impl Mul<Vec3D> for Vec3D {
    type Output = f64;

    fn mul(self, rhs: Vec3D) -> Self::Output {
        self * &rhs
    }
}

fn main() {
    let a = Vec3D{x: 1.0, y: 2.0, z: 3.0};
    let b = &Vec3D{x: 1.0, y: 2.0, z: 3.0};
    println!("{}", a * b);
    println!("{}", a * a);
}
```
The lint suggests removing the ampersand in the `Mul<Vec3D> for Vec3D` implementation. Doing that, however, leads to a warning about unconditional recursion, and, naturally, it causes the program to not function correctly. 

### Meta

- `cargo clippy -V`: clippy 0.1.56 (2d2bc94c 2021-08-15)
- `rustc -Vv`:
```
rustc 1.56.0-nightly (2d2bc94c8 2021-08-15)
binary: rustc
commit-hash: 2d2bc94c8c3aa778e191f80261c726e4777439f1
commit-date: 2021-08-15
host: x86_64-pc-windows-msvc
release: 1.56.0-nightly
LLVM version: 12.0.1
  ```
